### PR TITLE
gbm-kms: Support zwp_linux_dmabuf_unstable_v1

### PR DESCRIPTION
**Pull request by RAOF**
This is the new, improved way to do client buffer submission. Notably,
it features DRM modifier support, allowing clients to allocate fancy
driver-specific high-performance buffer formats.

Closes: #1663

### CODE REVIEW

File: src/platforms/gbm-kms/server/linux_dmabuf.cpp [hunk 1]
```diff
@@ -0,0 +1,846 @@
...
+    }
+
+    mir::geometry::Size size() const override
+    {
+        return size_;
+    }
+
+    MirPixelFormat pixel_format() const override
+    {
+        // There's no way to implement this corectly…
```
Comment (AlanGriffiths) [inline src/platforms/gbm-kms/server/linux_dmabuf.cpp:578]: Is there a discussion to be had?

File: src/platforms/gbm-kms/server/linux_dmabuf.cpp [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+    void destroy() override
+    {
+        destroy_wayland_object();
+    }
+
+    EGLDisplay const dpy;
+    std::shared_ptr<mg::EGLExtensions> const egl_extensions;
+    int32_t const width, height;
+    uint32_t const format_;
+    [[maybe_unused]]
```
Comment (AlanGriffiths) [inline src/platforms/gbm-kms/server/linux_dmabuf.cpp:196]: That's not going to compile on g++ 9.3.0 and earlier.

**Comment by AlanGriffiths** [comment]
18.04 still not happy! :disappointed:

File: src/platforms/gbm-kms/server/linux_dmabuf.cpp (hunk 1 above)
Comment (RAOF) [reply src/platforms/gbm-kms/server/linux_dmabuf.cpp:577]: Yes. We should lift something like [weston's pixel format descriptors](https://gitlab.freedesktop.org/wayland/weston/-/blob/master/libweston/pixel-formats.h) and replace the MirPixelFormat enum.

That's a change touching lots of code, though.

**Comment by RAOF** [comment]
Oh, arse. 18.04 has a too-old cmake to handle library dependencies on object libraries.

**Comment by RAOF** [comment]
Aaargh. And I can't get around it by sneakily installing the cmake snap in the PPA :cry:

**Review by AlanGriffiths — APPROVED** [review]
Looks sensible, and doesn't break stuff. (I'd love to be able to verify that it actually works, but don't have a way to do that.)

bors r+

**Comment by RAOF** [comment]
While you can't check if if fixes the Pi, you *can* check that it works on hardware; mesa will preferentially use the newly exported interface on Intel.

24 Sep 2020 19:03:51 Alan Griffiths <notifications@github.com>:

**Comment by AlanGriffiths** [comment]
> While you can't check if if fixes the Pi, you *can* check that it works on hardware; mesa will preferentially use the newly exported interface on Intel.

Well, Mir still works on Intel drivers. :)

**Comment by mariogrip** [comment]
Would this be possible to backport to 1.8? This would fix some (if not many) of our issues on arm devices.

**Pull request merged**